runtime/doc/lua-async.txt

```diff
@@ -0,0 +1,630 @@
+    coroutine.
+
+    Note a Task can be waited on via more than one waiter.
+
+    -- If a -- coroutine is awaiting on a Future object during cancellation,
+    the Future -- object will be cancelled.
+
+    Fields: ~
+      • {wait}            (`fun(self: vim.async.Task, callback_or_timeout: integer|fun(err?: any, ...: R...)?)`)
+                          See |Task:wait()|.
+      • {traceback}       (`fun(self: vim.async.Task, msg: string?): string`)
+                          See |Task:traceback()|.
+      • {raise_on_error}  (`fun(self: vim.async.Task): vim.async.Task`) See
+                          |Task:raise_on_error()|.
+      • {close}           (`fun(self: vim.async.Task, callback: fun(closed: boolean)?)`)
+                          See |Task:close()|.
+      • {status}          (`fun(self: vim.async.Task): 'running'|'suspended'|'normal'|'dead'?`)
+                          See |Task:status()|.
+
+*vim.vim.async.Future*
+    Future objects are used to bridge low-level callback-based code with
+    high-level async/await code.
+
+    Fields: ~
+      • {completed}  (`fun(self: vim.vim.async.Future): boolean`) See
+                     |Future:completed()|.
+      • {result}     (`fun(self: vim.vim.async.Future): boolean, any`) See
+                     |Future:result()|.
+      • {wait}       (`fun(self: vim.vim.async.Future, callback: fun(err?: any, ...: any))`)
+                     See |Future:wait()|.
+      • {complete}   (`fun(self: vim.vim.async.Future, err: string?, ...: any)`)
+                     See |Future:complete()|.
+
+
+await({...})                                               *vim.async.await()*
+    Asynchronous blocking wait
+
+    Example: >lua
+        local task = vim.async.run(function()
+           return 1, 'a'
+        end)
+
+        local task_fun = vim.async.async(function(arg)
+           return 2, 'b', arg
+        end)
+
+        vim.async.run(function()
+          do -- await a callback function
+            vim.async.await(1, vim.schedule)
+          end
+
+          do -- await a callback function (if function only has a callback argument)
+            vim.async.await(vim.schedule)
+          end
+
+          do -- await a task (new async context)
+            local n, s = vim.async.await(task)
+            assert(n == 1 and s == 'a')
+          end
+
+        end)
+<
+
+    Parameters: ~
+      • {...}  (`any`) see overloads
+
+    Overloads: ~
+      • `fun(func: (fun(callback: fun(...:R...)): vim.async.Closable?)): R...`
+      • `fun(argc: integer, func: (fun(...:T..., callback: fun(...:R...)): vim.async.Closable?), ...:T...): R...`
+      • `fun(task: vim.async.Task<R>): R...`
+
+event()                                                    *vim.async.event()*
+    Create a new event
+
+    An event can signal to multiple listeners to resume execution The event
+    can be set from a non-async context. >lua
+         local event = vim.async.event()
```

Comment:
Is this similar to Go channels? 

This seems like something that could wait until a later phase. Do we need this in the first phase?

Reply:
Personally, I think it’s better to keep the API surface small for the initial release, limiting it to vim.async.{run, all, any, await}. This would make it easier to get merged.

Reply:
`event`, `queue`, `future` and `semaphore` were ripped directly from [nvim-nio](https://github.com/nvim-neotest/nvim-nio).

`event` is used to implement `semaphore` which I already have 3 downstream users waiting for: nvim-treesitter, gitsigns and vim.pack.

I'm happy to hide `event`, `future` and `queue` from the API (and keep hidden), but I really want to keep semaphore in this PR since it provides a very nice way to control throughput of jobs.

Reply:
Ok that sounds fair. We can always expose more stuff later, but it will help a lot to clearly signal the "core concepts" vs the "low level concepts".